src/app/modules/form/public-form/public-form.controller.ts

```diff
@@ -350,3 +364,92 @@ export const handleGetPublicForm: RequestHandler<
       return new UnreachableCaseError(authType)
   }
 }
+
+/**
+ * NOTE: This is exported only for testing
+ * Generates redirect URL to Official SingPass/CorpPass log in page
+ * @param isPersistentLogin whether the client wants to have their login information stored
+ * @returns 200 with the redirect url when the user authenticates successfully
+ * @returns 400 when there is an error on the authType of the form
+ * @returns 400 when the eServiceId of the form does not exist
+ * @returns 404 when form with given ID does not exist
+ * @returns 500 when database error occurs
+ * @returns 500 when the redirect url could not be created
+ * @returns 500 when the redirect feature is not enabled
+ */
+export const _handleFormAuthRedirect: RequestHandler<
+  { formId: string },
+  RedirectUrlDto | ErrorDto,
+  unknown,
+  Query & { isPersistentLogin: boolean }
+> = (req, res) => {
+  const { formId } = req.params
+  const { isPersistentLogin } = req.query
+  const logMeta = {
+    action: 'handleFormAuthRedirect',
+    ...createReqMeta(req),
+    formId,
+  }
+  // NOTE: Using retrieveFullForm instead of retrieveForm to ensure authType always exists
+  return FormService.retrieveFullFormById(formId)
+    .andThen((form) => {
+      switch (form.authType) {
+        case AuthType.MyInfo:
+          return validateMyInfoForm(form).andThen((form) =>
+            MyInfoFactory.createRedirectURL({
+              formEsrvcId: form.esrvcId,
+              formId,
+              requestedAttributes: form.getUniqueMyInfoAttrs(),
+            }),
+          )
+        case AuthType.SP:
+        case AuthType.CP: {
+          // NOTE: Persistent login is only set (and relevant) when the authType is SP.
+          // If authType is not SP, assume that it was set erroneously and default it to false
+          return validateSpcpForm(form).andThen((form) => {
+            const target = `/${formId},${
+              // Need to cast to boolean because undefined is allowed as a valid value
+              // We are not following corppass's official spec for
+              // the target parameter
+              form.authType === AuthType.SP ? !!isPersistentLogin : false
+            }`
+            return SpcpFactory.createRedirectUrl(
+              form.authType,
+              target,
+              form.esrvcId,
+            )
+          })
+        }
+        // NOTE: Only MyInfo and SPCP should have redirects as the point of a redirect is
+        // to provide auth for users from a third party
+        default:
+          return err<never, AuthTypeMismatchError>(
```

Comment:
am I right to assume that providing the generic types is necessary so that `mapErr` and `map` have the correct types?

Reply:
yeap! in general, we actually need to hint when we construct `ok` or `err` directly. this is because they are type constructors and take 2 arguments, 1 of which can never occur.